alibi/explainers/counterfactual_tf2.py

```diff
@@ -0,0 +1,1028 @@
+
+                # save search results and log to TensorBoard
+                write_summary = self.debug and self.step % self.summary_freq == 0
+                current_state = (self.step, lam, to_numpy_arr(self.cf), to_numpy_arr(cf_prediction))
+                if constraint_satisfied:
+                    cf_found[lam_step] += 1
+                    self._update_search_result(*current_state)
+                else:
+                    if write_summary:
+                        self._collect_step_data(*current_state)
+
+                if write_summary:
+                    self._write_tb(self.step, lam, 0, 0, cf_found, prefix='lambda_sweep/')
+                self.step += 1
+            self.lam_step += 1
+
+        self._reset_step()
+
+        logger.debug(f"Counterfactuals found: {cf_found}")
+
+        # determine upper and lower bounds given the solutions found
+        lb_idx_arr = np.where(cf_found > 0)[0]
+        if len(lb_idx_arr) < 2:
+            logger.exception("Could not find lower bound for lambda, try decreasing lam_init!")
+            raise CounterfactualError("Could not find lower bound for lambda, try decreasing lam_init!")
+        lam_lb = lams[lb_idx_arr[1]]
+
+        ub_idx_arr = np.where(cf_found == 0)[0]
+        if len(ub_idx_arr) == 1:
+            logger.warning(
+                f"Could not find upper bound for lambda where no solutions found. "
+                f"Setting upper bound to lam_init={lams[0]}")
+            ub_idx = 0
+        else:
+            ub_idx = ub_idx_arr[-1]
+        lam_ub = lams[ub_idx]
+
+        logger.debug(f"Found upper and lower bounds for lambda: {lam_ub}, {lam_lb}")
+        assert lam_lb - lam_ub > -1e-3
+
+        # re-initialise response s.t. 'cf' and 'all' fields contain only main search resuts
+        sweep_results = deepcopy(self.search_results)
+        self._initialise_response()
+        self.search_results['lambda_sweep']['all'] = sweep_results['all']
+        self.search_results['lambda_sweep']['cf'] = sweep_results['cf']
+
+        return {'lb': lam_lb, 'ub': lam_ub, 'midpoint': 0.5*(lam_ub + lam_lb)}
+
+    def _cf_step(self, lam: float):
+        """
+        Runs a gradient descent step and updates current solution.
+        """
+        # TODO: ALEX: DOCSTRING
+
+        gradients = self._get_gradients(lam)
+        self._apply_gradients(gradients)
+
+    def _get_gradients(self, lam: float) -> List[tf.Tensor]:
+        """
+        Calculates the gradients of the counterfactual loss.
+        """
+        # TODO: ALEX: DOCSTRING
+
+        with tf.GradientTape() as tape:
+            prediction = self.get_cf_prediction(self.target_class, self.cf)
+            loss = self.loss(
+                instance=self.instance,
+                cf=self.cf,
+                lam=lam,
+                pred_probas=prediction,
+                target_probas=self.target_proba,
+            )
+        gradients = tape.gradient(loss, [self.cf])
+
+        return gradients
+
```

Comment:
Not urgent but need to keep in mind that potentially numerical gradients could be added as well.